### ISSUE TITLE
queue-output write only maximum values within a configurable period

### ISSUE THREAD

**Issue by namdre** [eclipse/sumo#9233]
via option **--queue-output.max-period**

for each time window of max-period length, write only the timestep where the queue had maximal length

**Comment by mennowo** [comment]
a consideration could be to log only temporary maximum queue lengths, eg. via an option --queue-output.max-queue-threshold or similar.

the idea being, that you'd be interested mostly in the moments where the queue reaches some maximum, before shrinking again, for example because of a green traffic light

this could be implemented by holding a current maximum; then if the queue length starts decreasing, once it goes below the threshold, log the max value that occurred and restart the measurement. this way you'd have only max queue lengths as output, which is data that is compact and easy to process

I'd be happy to implement this (in MSQueueExport.cpp?), but I might need some pointers. eg. how/where to store a buffer of data holding the current maximum queue length. and then, maybe some pointer on how/where to implement the command line option

**Comment by namdre** [comment]
For adding and using options, see the commits for #9232
You can store temporary data as static members of MSQueue-Export but then have to make sure gets cleaned up on reload somewhere in MSNet::clearAll.

Adding an option for a threshold could also be combined independent of all other options: only log queue lengths when above this threshold. 

 Your approach of using the threshold-traversal to reset the maximum value might be prone to frequent output if the queue length happens to hover around that threshold value. Logging the maximum value in a fixed time window (as suggested in the original comment) might be more reliable in that regard. 

If it is combined with a threshold you would still only see time windows that are of interest. The main difference to your approach would be that you would get multiple outputs if the queue length stays above the threshold for multiple intervals.

**Comment by mennowo** [comment]
The way I do this in Python (post-processing the current output) is:
- if current queue > stored queue: set stored value to current value
- if current queue <= stored queue minus threshold (ie. 5 meters), log the stored value, then set stored to 0
This results in a list of points in time where the queue was at a momentary max.
Of course, there may be cases where this will result in frequent output, but with a correct threshold, this will usually not occur.
Maybe my use case is rather specific: I wish not just to filter maximum queues, but get all instances where the queue reached a (smaller of bigger) peak before shrinking. Thus, even a single car that leaves should yield a result.

**Comment by mennowo** [comment]
Is there a way I can get all edges/lanes at the start of the simulation inside `MSQueueExport`, so I can use this subsequently during timesteps? I'm just trying this out, and it would be nice to be able to create a map of edges and hold data there.

**Comment by namdre** [comment]
see function ::writeEdge which shows how to iterate over all simulation edges and lanes.

**Comment by mennowo** [comment]
Yes, but in order to have a map or array with all edges to begin with so I can store data for them, it would make more sense to set that up once at initialization. Unless that is illogical in some way.

My goal would be to only log once the queue shrinks below a given threshold. That requires a somewhat different approach than the one currently implemented.

**Comment by namdre** [comment]
I misunderstood your question. The easiest way to handle initialization is to check whether your new static data container has size 0 or not and (and then conditionally init it with the retrieve-all-edges code in `writeEdge`)
